File: app/src/main/java/com/botob/ulisten2/media/AbstractMedia.java
package com.botob.ulisten2.media;

import android.os.Parcel;
import android.os.Parcelable;

import com.botob.ulisten2.notification.NotificationData;

import java.lang.reflect.InvocationTargetException;
import java.util.Objects;

/**
 * Created by boris on 12/9/14.
 */
public abstract class AbstractMedia implements Media {
    /**
     * The parcelable creator.
     */
    public static final Parcelable.Creator<Media> CREATOR = new Parcelable.Creator<Media>() {
        public Media createFromParcel(Parcel source) {
            Media media = null;
            try {
                media = MediaApp.valueOf(source.readString()).instantiate(source);
            } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException | InstantiationException e) {
                e.printStackTrace();
            }
            return media;
        }

        public AbstractMedia[] newArray(int size) {
            return new AbstractMedia[size];
        }
    };

    private static final String[] AD_KEYWORDS = new String[]{
            "advertisement"
    };

    private NotificationData mNotificationData;
    private String mTitle;
    private String mAlbum;
    private String mArtist;
    private long mBroadcastTime;
    private String mPackageName;

    public AbstractMedia(NotificationData notificationData) {
        mNotificationData = notificationData;
        mTitle = fetchTitle(notificationData);
        mAlbum = fetchAlbum(notificationData);
        mArtist = fetchArtist(notificationData);
        mBroadcastTime = mNotificationData.postTime;
        mPackageName = mNotificationData.packageName;
    }

    public AbstractMedia(Parcel parcel) {
        mTitle = parcel.readString();
        mAlbum = parcel.readString();
        mArtist = parcel.readString();
        mBroadcastTime = parcel.readLong();
        mPackageName = parcel.readString();
    }

    public AbstractMedia(final String title, final String album, final String artist,
                         final long timestamp, final String packageName) {
        mTitle = title;
        mAlbum = album;
        mArtist = artist;
        mBroadcastTime = timestamp;
        mPackageName = packageName;
    }

    @Override
    public boolean equals(Object object) {
        final boolean result;
        if (object == null || !getClass().equals(object.getClass())) {
            result = false;
        } else {
            Media otherMedia = (Media) object;
            result = getTitle().equals(otherMedia.getTitle())
                    && getAlbum().equals(otherMedia.getAlbum())
                    && getArtist().equals(otherMedia.getArtist());
        }
        return result;
    }

    @Override
    public int hashCode() {
        return Objects.hash(getTitle(), getAlbum(), getArtist());
    }

    public boolean isRelevant() {
        return getTitle() != null && getArtist() != null && !isAd();
    }

    private boolean isAd() {
        boolean result = false;
        for (int i = 0; i < AD_KEYWORDS.length && !result; i++) {
            result = AD_KEYWORDS[i].equalsIgnoreCase(getTitle());
        }
        return result;
    }

    public int describeContents() {
        return 0;
    }

    public void writeToParcel(Parcel out, int flags) {
        out.writeString(mTitle);
        out.writeString(mAlbum);
        out.writeString(mArtist);
    }

    public String getTitle() {
        return mTitle != null ? mTitle : "";
    }

    public String getAlbum() {
        return mAlbum != null ? mAlbum : "";
    }

    public String getArtist() {
        return mArtist != null ? mArtist : "";
    }

    @Override
    public long getBroadcastTime() {
        return mNotificationData != null ? mNotificationData.postTime : System.currentTimeMillis();
    }

    @Override
    public String getPackageName() {
        return mNotificationData.packageName;
    }

    protected abstract String fetchTitle(NotificationData notificationData);

    protected abstract String fetchAlbum(NotificationData notificationData);

    protected abstract String fetchArtist(NotificationData notificationData);
}

File: app/src/main/java/com/botob/ulisten2/media/AbstractMedia.kt
package com.botob.ulisten2.media

import android.os.Parcel
import android.os.Parcelable
import com.botob.ulisten2.notification.NotificationData
import java.lang.reflect.InvocationTargetException
import java.util.*

/**
 * Created by boris on 12/9/14.
 */
abstract class AbstractMedia : Media {
    companion object {
        /**
         * The parcelable creator.
         */
        val CREATOR: Parcelable.Creator<Media> = object : Parcelable.Creator<Media> {
            override fun createFromParcel(source: Parcel): Media? {
                var media: Media? = null
                try {
                    media = MediaApp.valueOf(source.readString()!!).instantiate(source)
                } catch (e: NoSuchMethodException) {
                    e.printStackTrace()
                } catch (e: IllegalAccessException) {
                    e.printStackTrace()
                } catch (e: InvocationTargetException) {
                    e.printStackTrace()
                } catch (e: InstantiationException) {
                    e.printStackTrace()
                }
                return media
            }

            override fun newArray(size: Int): Array<AbstractMedia?> {
                return arrayOfNulls(size)
            }
        }
        private val AD_KEYWORDS = arrayOf(
            "advertisement"
        )
    }

    private var _notificationData: NotificationData? = null
    private var _title: String? = null
    private var _album: String? = null
    private var _artist: String? = null
    private var _broadcastTime: Long
    private var _packageName: String?

    constructor(notificationData: NotificationData) {
        _notificationData = notificationData
        _broadcastTime = _notificationData!!.postTime
        _packageName = _notificationData!!.packageName
    }

    constructor(parcel: Parcel) {
        _title = parcel.readString()
        _album = parcel.readString()
        _artist = parcel.readString()
        _broadcastTime = parcel.readLong()
        _packageName = parcel.readString()
    }

    constructor(
        title: String?, album: String?, artist: String?,
        timestamp: Long, packageName: String?
    ) {
        _title = title
        _album = album
        _artist = artist
        _broadcastTime = timestamp
        _packageName = packageName
    }

    override fun equals(other: Any?): Boolean {
        val result: Boolean = if (other == null || javaClass != other.javaClass) {
            false
        } else {
            val otherMedia = other as Media
            title == otherMedia.title && album == otherMedia.album && artist == otherMedia.artist
        }
        return result
    }

    override fun hashCode(): Int {
        return Objects.hash(title, album, artist)
    }

    override val isRelevant: Boolean
        get() = title != null && artist != null && !isAd
    override val isAd: Boolean
        get() {
            var result = false
            var i = 0
            while (i < AD_KEYWORDS.size && !result) {
                result = AD_KEYWORDS[i].equals(title, ignoreCase = true)
                i++
            }
            return result
        }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(out: Parcel, flags: Int) {
        out.writeString(title)
        out.writeString(album)
        out.writeString(artist)
    }

    override val title: String?
        get() = _title ?: fetchTitle(_notificationData)

    override val album: String
        get() = _album ?: fetchAlbum(_notificationData)

    override val artist: String?
        get() = _artist ?: fetchArtist(_notificationData)

    override val broadcastTime: Long
        get() = if (_notificationData != null) _notificationData!!.postTime else System.currentTimeMillis()

    override val packageName: String?
        get() = _notificationData!!.packageName
}